cinn/ir/schedule_desc.cc

```diff
@@ -0,0 +1,577 @@
+#define SET_DESC_SINGLE_ITEM(index, built_type, proto_type, proto_field)   \
+  case index:                                                              \
+    attr_proto->set_dtype(proto::ScheduleDesc_Attr_DataType_##proto_type); \
+    attr_proto->set_##proto_field(absl::get<built_type>(attr));            \
+    break;
+
+#define SET_DESC_REPEATED_ITEM(index, built_type, proto_type, proto_field) \
+  case index: {                                                            \
+    attr_proto->set_dtype(proto::ScheduleDesc_Attr_DataType_##proto_type); \
+    const auto& values = absl::get<built_type>(attr);                      \
+    attr_proto->mutable_##proto_field()->Reserve(values.size());           \
+    *attr_proto->mutable_##proto_field() = {values.begin(), values.end()}; \
+    break;                                                                 \
+  }
+
+  switch (attr.index()) {
+    SET_DESC_SINGLE_ITEM(0, bool, BOOLEAN, b);
+    SET_DESC_SINGLE_ITEM(1, float, FLOAT, f);
+    SET_DESC_SINGLE_ITEM(2, int, INT, i);
+    SET_DESC_SINGLE_ITEM(3, std::string, STRING, s);
+    SET_DESC_REPEATED_ITEM(4, std::vector<bool>, BOOLEANS, bools);
+    SET_DESC_REPEATED_ITEM(5, std::vector<int>, INTS, ints);
+    SET_DESC_REPEATED_ITEM(6, std::vector<float>, FLOATS, floats);
+    SET_DESC_REPEATED_ITEM(7, std::vector<std::string>, STRINGS, strings);
+    default:
+      LOG(FATAL) << "Invalid index:" << attr.index();
+  }
+
+#undef SET_DESC_SINGLE_ITEM
+#undef SET_DESC_REPEATED_ITEM
+}
+
+utils::Attribute AttrProtoToVariant(const proto::ScheduleDesc_Attr& attr) {
+  utils::Attribute value;
+#define PARSE_DESC_SINGLE_ITEM(proto_type, proto_field, built_type) \
+  case proto::ScheduleDesc_Attr_DataType_##proto_type:              \
+    value = built_type(attr.proto_field());                         \
+    break;
+
+#define PARSE_DESC_REPEATED_ITEM(proto_type, proto_field, built_type)           \
+  case proto::ScheduleDesc_Attr_DataType_##proto_type:                          \
+    value = built_type({attr.proto_field().begin(), attr.proto_field().end()}); \
+    break;
+
+  switch (attr.dtype()) {
+    PARSE_DESC_SINGLE_ITEM(BOOLEAN, b, bool);
+    PARSE_DESC_SINGLE_ITEM(INT, i, int);
+    PARSE_DESC_SINGLE_ITEM(FLOAT, f, float);
+    PARSE_DESC_SINGLE_ITEM(STRING, s, std::string);
+    PARSE_DESC_REPEATED_ITEM(BOOLEANS, bools, std::vector<bool>);
+    PARSE_DESC_REPEATED_ITEM(INTS, ints, std::vector<int>);
+    PARSE_DESC_REPEATED_ITEM(FLOATS, floats, std::vector<float>);
+    PARSE_DESC_REPEATED_ITEM(STRINGS, strings, std::vector<std::string>);
+    default:
+      LOG(FATAL) << "Invalid type:" << attr.DebugString();
+  }
+
+#undef PARSE_DESC_SINGLE_ITEM
+#undef PARSE_DESC_REPEATED_ITEM
+  return value;
+}
+
+// Expr hash functor, presents how to hash an Expr
+struct ExprHash {
+  size_t operator()(const Expr& e) const { return std::hash<IrNode*>()(e.ptr()); }
+};
+// Expr equal functor, presents whether a Expr pair is equal
+struct ExprEqual {
```

Comment:
Maybe renaming to ExprPointerEqual? Since it is just comparing pointer, not object equal.